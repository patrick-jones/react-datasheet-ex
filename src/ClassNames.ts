export const READ_ONLY_CELL = 'cell read-only';
export const ACTION_CELL = 'rdx-action-cell';

export const HEADER_CELL = 'rdx-header-cell';
export const HEADER_CELL_LEFT = 'rdx-header-cell--left';
export const HEADER_CELL_CENTER = 'rdx-header-cell--center';
export const HEADER_CELL_RIGHT = 'rdx-header-cell--right';

export const HEADER_CELL_CONTAINER = 'rdx-header-cell-container';
export const HEADER_CELL_TITLE = 'rdx-header-cell-title';

export const HEADER_DRAG_SOURCE = 'rdx-header-drag-source';
export const HEADER_DROP_TARGET = 'rdx-header-drop-target';

export const ROW = 'rdx-row';
export const ROW_DROP_TARGET = 'rdx-row-drop-target';
export const ROW_DRAG_SOURCE = 'rdx-row-drag-source';

export const RESIZING_WIDTH = 'rdx-ex-resizing-width';
export const DATA_GRID_RESIZABLE = 'rdx-data-grid-resizable';
export const WIDTH_RESIZE_HANDLE = 'rdx-width-resize-handle';

export const SCROLL_PANEL = 'rds-ex-scroll-panel';
