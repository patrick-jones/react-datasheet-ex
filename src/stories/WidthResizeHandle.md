# Resize Handle

This thin vertical bar actually implements all of the resizable logic;
the other components are built on top of it. It requires the `ResizableProps`
and should be placed within your header.


<!-- STORY -->


```jsx
import {HeaderContainer, HeaderTitle} from 'react-datasheet-ex';
import {WidthResizeHandle} from 'react-datasheet-ex/resize';

class MyClass extends Component {
  
  headerRenderer = (headerProps) => (
    <HeaderContainer>
      <HeaderTitle {...headerProps} />
      <WidthResizeHandle
        index={headerProps.col}
        onResizing={this.props.actions.headerResizing}
        onResizeEnd={this.props.actions.headerResizeEnd}
      />
    </HeaderContainer>
  );
    
  // ...rest of class...
}
```
