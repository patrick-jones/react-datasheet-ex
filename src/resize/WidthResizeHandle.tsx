import React, {PureComponent, ReactNode} from 'react';

import {RESIZING_WIDTH, WIDTH_RESIZE_HANDLE} from '../ClassNames';
import {ResizableHeaderProps} from './interfaces';


const initialState = {
  initialOffset: 0,
};


export default class WidthResizeHandle
  extends PureComponent<ResizableHeaderProps, typeof initialState> {

  static defaultProps = {
    cellTag: 'TH',
    minSize: 17,
  };

  state = initialState;

  elementRef = React.createRef<HTMLDivElement>();

  originalWidth = () => {
    const th = header(this.elementRef.current, this.props.cellTag);
    return th ? th.clientWidth : 0;
  };

  handleMouseDown = (e: React.MouseEvent<HTMLElement>) => {
    const {index, onResizeStart} = this.props;
    const originalSize = this.originalWidth();
    const initialOffset = originalSize - e.pageX;

    this.setState({initialOffset});
    window.addEventListener('mouseup', this.handleMouseUp);
    window.addEventListener('mousemove', this.handleMouseMoved);
    document.body.classList.add(RESIZING_WIDTH);
    onResizeStart && onResizeStart(index, originalSize);
  };

  handleMouseMoved = (e: MouseEvent) => {
    const {index, onResizing, minSize} = this.props;
    if (onResizing) {
      const {initialOffset} = this.state;
      const requestedSize = initialOffset + e.pageX;
      if (requestedSize >= minSize) {
        onResizing(index, requestedSize);
      }
    }
  };

  handleMouseUp = (e: MouseEvent) => {
    window.removeEventListener('mouseup', this.handleMouseUp);
    window.removeEventListener('mousemove', this.handleMouseMoved);
    document.body.classList.remove(RESIZING_WIDTH);
    const {index, onResizeEnd, minSize} = this.props;
    const {initialOffset} = this.state;
    this.setState(initialState);
    onResizeEnd(index, Math.max(initialOffset + e.pageX, minSize));
  };

  render(): ReactNode {
    const {className, style} = this.props;

    return (
      <span
        ref={this.elementRef}
        className={className ? `${WIDTH_RESIZE_HANDLE} ${className}` : WIDTH_RESIZE_HANDLE}
        style={style}
        onMouseDown={this.handleMouseDown}
      >
        {this.props.children}
      </span>
    );
  }
}


export function header(el: HTMLElement | null, selector: string): Element | null {
  if (!el) {
    return null;
  }

  const upper = selector.toUpperCase();
  if (el.closest) {
    return el.closest(upper);
  }


  let parent = el.parentElement;
  while (parent && parent.nodeType === Node.ELEMENT_NODE) {
    if (parent.tagName === upper) {
      return parent;
    }
    parent = parent.parentElement;
  }

  return null;
}
