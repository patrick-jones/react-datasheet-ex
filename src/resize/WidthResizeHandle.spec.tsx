import React from 'react';
import 'jest';
import {mount, ReactWrapper} from 'enzyme';
import toJson from 'enzyme-to-json';
import WidthResizeHandle, {header} from './WidthResizeHandle';
import {WIDTH_RESIZE_HANDLE} from '../ClassNames';


describe('WidthResizeHandle', () => {
  const start = jest.fn();
  const during = jest.fn();
  const end = jest.fn();
  let wrapper: ReactWrapper<HTMLElement>;
  let map: {[key: string]: (...args: any[]) => void} = {};
  let ael = window.addEventListener;

  beforeEach(() => {
    start.mockClear();
    during.mockClear();
    end.mockClear();
    wrapper = mount(
      <table>
        <thead>
        <tr>
          <th>
            <WidthResizeHandle
              index={0}
              minSize={20}
              onResizeStart={start}
              onResizing={during}
              onResizeEnd={end}
            />
          </th>
        </tr>
        </thead>
      </table>
    );

    // got to fake the funk since jsdom doesn't really do layout
    const th = wrapper.find('th').first().getDOMNode();
    Object.defineProperty(th, 'clientWidth', {value: 100, writable: false});

    // got to fake document events because enzyme won't do it for you
    // https://github.com/airbnb/enzyme/issues/426#issuecomment-228601631
    ael = window.addEventListener;
    window.addEventListener = jest.fn((event, cb) => {
      map[event] = cb;
    });
  });

  afterEach(() => {
    wrapper.unmount();
    window.addEventListener = ael;
    map = {};
  });

  it('Should render', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should handle interaction', () => {
    wrapper.find(`.${WIDTH_RESIZE_HANDLE}`).simulate('mouseDown', {pageX: 198});

    map.mousemove({pageX: 148});
    map.mousemove({pageX: 118});
    map.mousemove({pageX: 48}); // skipped, < minSize
    map.mouseup({pageX: 158});

    expect(start.mock.calls.length).toBe(1);
    expect(start.mock.calls[0][0]).toBe(0);
    expect(start.mock.calls[0][1]).toBe(100);

    expect(during.mock.calls.length).toBe(2);
    expect(during.mock.calls[0][0]).toBe(0);
    expect(during.mock.calls[0][1]).toBe(50);
    expect(during.mock.calls[1][1]).toBe(20); // minSize

    expect(end.mock.calls.length).toBe(1);
    expect(end.mock.calls[0][0]).toBe(0);
    expect(end.mock.calls[0][1]).toBe(60);
  });

  it('Should return `originalWidth` of zero if no parent is found', () => {
    const handler = jest.fn();
    const wrapper = mount(
      <table>
        <thead>
        <tr>
          <th>
            <WidthResizeHandle
              index={0}
              onResizeStart={handler}
              onResizing={handler}
              onResizeEnd={handler}
              cellTag='P'
            />
          </th>
        </tr>
        </thead>
      </table>
    );

    const th = wrapper.find('th').first().getDOMNode();
    Object.defineProperty(th, 'clientWidth', {value: 100, writable: false});

    const inst = wrapper.find('WidthResizeHandle').instance() as WidthResizeHandle;
    expect(inst.originalWidth()).toBe(0);
  });

});

describe('WidthResizeHandle.closest', () => {

  let inner: string;
  let th: HTMLElement;
  let span: HTMLElement;

  beforeEach(() => {
    inner = document.body.innerHTML;
    document.body.innerHTML =
      `<table><thead><th id='p'><div><span id='c'>X</span></div></th></thead></table>`;

    th = document.getElementById('p')!;
    span = document.getElementById('c')!;
    Object.defineProperty(span, 'closest', {value: undefined, writable: false});
  });

  afterEach(() => {
    document.body.innerHTML = inner;
  });

  it('Should find the parent `th` event if `closest` is not available (curse you IE)', () => {
    const parent = header(span, 'th');
    expect(parent).toBe(th);
  });

  it('Should return a null if the selector is not found', () => {
    const parent = header(span, 'P');
    expect(parent).toBeNull();
  });

  it('Should return null if passed a null element', () => {
    const parent = header(null, 'TH');
    expect(parent).toBeNull();
  });
});
