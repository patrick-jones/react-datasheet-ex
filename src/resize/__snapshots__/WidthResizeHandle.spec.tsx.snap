// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidthResizeHandle Should render 1`] = `
<table>
  <thead>
    <tr>
      <th>
        <WidthResizeHandle
          cellTag="TH"
          index={0}
          minSize={20}
          onResizeEnd={[MockFunction]}
          onResizeStart={[MockFunction]}
          onResizing={[MockFunction]}
        >
          <span
            className="rdx-width-resize-handle"
            onMouseDown={[Function]}
          />
        </WidthResizeHandle>
      </th>
    </tr>
  </thead>
</table>
`;
